clist/clist.go

```diff
@@ -152,48 +206,55 @@ func (l *CList) Init() *CList {
 func New() *CList { return new(CList).Init() }
 
 func (l *CList) Len() int {
-	l.mtx.Lock()
-	defer l.mtx.Unlock()
+	l.mtx.RLock()
+	defer l.mtx.RUnlock()
+
 	return l.len
 }
 
 func (l *CList) Front() *CElement {
-	l.mtx.Lock()
-	defer l.mtx.Unlock()
+	l.mtx.RLock()
+	defer l.mtx.RUnlock()
+
 	return l.head
 }
 
 func (l *CList) FrontWait() *CElement {
 	for {
-		l.mtx.Lock()
+		l.mtx.RLock()
 		head := l.head
 		wg := l.wg
-		l.mtx.Unlock()
-		if head == nil {
-			wg.Wait()
-		} else {
+		l.mtx.RUnlock()
+
+		if head != nil {
 			return head
 		}
+		wg.Wait()
+		// l.head doesn't necessarily exist here.
```

Comment:
Perhaps we could add some more clarity to this comment?
Could we comment right before the ```for {```
```go
// Loop until the head is non-nil else wait and try again
```
and then we won't need the comment right here.

However, I believe that this is trying to reimplement the logic that sync.Cond is meant
for, moreover without the runtime polls. What do you think of us instead using sync.Cond?

Reply:
See https://github.com/tendermint/tmlibs/pull/113/files#diff-0326c2957f72e5e4cf728fcfdfde3521R31
I like the comment suggestion, will do.

Reply:
Thanks Jae! I understood though that head == nil at that point, I just didn't see a need for that comment hence
requested that we just make a single comment instead of the current one that has been added.